File: android/app/src/main/java/app/headstart/data/PairRepository.kt
package app.headstart.data

import app.headstart.core.Callables
import app.headstart.core.HeadstartError
import app.headstart.core.snapshotFlow
import app.headstart.core.str
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

/** The 6-character alphabet from CLIENT_CONTRACT.md — no I, O, 0 or 1. */
object InviteCode {
    const val ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
    private const val LENGTH = 6

    /** Upper-cases, strips anything that is not in the alphabet, and checks the length. */
    fun normalize(raw: String?): String? {
        val cleaned = raw.orEmpty().uppercase().filter { it in ALPHABET }
        return if (cleaned.length == LENGTH) cleaned else null
    }

    /** Addendum §N: the deep-link scheme is `headstart` everywhere. The pre-rename scheme appears nowhere. */
    fun link(code: String): String = "headstart://pair/$code"

    /** The bare code leads, so manual entry always works even where the link cannot open. */
    fun shareText(code: String): String =
        "Pair with me on Headstart. Code: $code — ${link(code)}"

    /** Pulls a code out of `headstart://pair/{code}` or `https://.../pair/{code}`. */
    fun fromDeepLink(uri: String?): String? {
        val text = uri ?: return null
        val marker = "/pair/"
        val at = text.indexOf(marker)
        if (at < 0) return null
        return normalize(text.substring(at + marker.length).substringBefore('?').substringBefore('/'))
    }
}

data class CreatedPair(val pairId: String, val inviteCode: String)

class PairRepository(
    private val db: FirebaseFirestore,
    private val callables: Callables,
) {
    /**
     * All pairs this user belongs to. Queried by `array-contains` alone and filtered in
     * Kotlin: adding `status == active` to the query needs the composite index the backend
     * declares (addendum §M), and this way the client works whether or not it exists yet.
     * A user is only ever in a handful of pair documents.
     */
    fun pairsFor(uid: String): Flow<List<PairInfo>> =
        db.collection("pairs")
            .whereArrayContains("members", uid)
            .snapshotFlow()
            .map { snap -> snap.documents.mapNotNull { pairFrom(it.id, it.data) } }

    fun activePair(uid: String): Flow<PairInfo?> =
        pairsFor(uid).map { list -> list.firstOrNull { it.isActive } }

    /** The invite this user created and nobody has accepted yet — drives PairInvite. */
    fun pendingInvite(uid: String): Flow<PairInfo?> =
        pairsFor(uid).map { list -> list.firstOrNull { !it.isActive && it.createdBy == uid } }

    /** One pair document, live. `memberNames` lands here when the other side signs in. */
    fun pair(pairId: String): Flow<PairInfo?> =
        db.collection("pairs").document(pairId).snapshotFlow()
            .map { snap -> snap?.let { pairFrom(it.id, it.data) } }

    suspend fun createPair(): CreatedPair {
        val res = callables.call("createPair")
        val pairId = res.str("pairId") ?: throw HeadstartError.Unknown("createPair-no-id")
        val code = res.str("inviteCode") ?: throw HeadstartError.Unknown("createPair-no-code")
        return CreatedPair(pairId, code)
    }

    suspend fun acceptPair(rawCode: String): String {
        val code = InviteCode.normalize(rawCode) ?: throw HeadstartError.BadCode
        val res = callables.call("acceptPair", mapOf("code" to code))
        return res.str("pairId") ?: throw HeadstartError.Unknown("acceptPair-no-id")
    }

    suspend fun revokePair(pairId: String) {
        callables.call("revokePair", mapOf("pairId" to pairId))
    }
}

File: ios/Headstart/Data/PairRepository.swift
// ios/Headstart/Data/PairRepository.swift
//
// Owns the single pair this user belongs to. Every mutation goes through a callable
// (`createPair`, `acceptPair`, `revokePair`) — the client never writes `pairs/{id}` — and the
// live read is a snapshot listener behind an `AsyncStream` (decision D8).
//
// Invite codes: the client only DISPLAYS a code the server minted and ACCEPTS one a person
// typed. It never generates one — CLIENT_CONTRACT.md line 12 fixes the alphabet server-side,
// and `PhoneNumber.sanitizeInviteCode` is the entry filter.

import Foundation
import FirebaseFirestore

@MainActor
public final class PairRepository: ObservableObject {

    /// The active pair if there is one; otherwise the PENDING invite this user created, so
    /// `PairInviteView` can keep showing the same code across app launches instead of
    /// minting a second one.
    @Published public private(set) var pair: Pair?

    /// False until the first snapshot lands. A screen that branches on `pair == nil` before
    /// this is true will flash the "not paired" state at every launch.
    @Published public private(set) var hasLoaded = false

    private let db: Firestore
    private let callables: Callables
    private var task: Task<Void, Never>?

    public init(db: Firestore, callables: Callables) {
        self.db = db
        self.callables = callables
    }

    /// ADDENDUM §M — a client finds its own pair with `members array-contains uid`. The
    /// `status` filter is applied HERE rather than in the query on purpose: as a single-field
    /// query this needs no composite index, and we also want the pending row (see `pair`).
    public func observe(uid: String) {
        task?.cancel()
        hasLoaded = false
        let stream = db.collection("pairs")
            .whereField("members", arrayContains: uid)
            .modelsStream { Pair(id: $0, data: $1) }
        task = Task { [weak self] in
            for await pairs in stream {
                guard let self else { return }
                self.pair = Self.pick(from: pairs, uid: uid)
                self.hasLoaded = true
            }
        }
    }

    /// Active always wins; a pending row only counts when this user is the one who created
    /// it, because a pending row created by someone else is just an invite they have not
    /// accepted and must not look like "you are paired".
    static func pick(from pairs: [Pair], uid: String) -> Pair? {
        pairs.first(where: \.isActive)
            ?? pairs.first { $0.isPending && $0.createdBy == uid }
    }

    public func stop() {
        task?.cancel()
        task = nil
        pair = nil
        hasLoaded = false
    }

    // MARK: - Callables

    /// Returns the six-character invite code to show on `PairInviteView`.
    public func create() async throws -> CreatePairResponse {
        try await callables.createPair()
    }

    /// `code` is sanitised here as well as in the field, so a caller that skipped the field
    /// (a deep link, `headstart://pair/K7M2QP`) cannot send something the server will bounce.
    /// Errors the UI must be ready for: `.badCode`, `.ownCode`.
    @discardableResult
    public func accept(code: String) async throws -> String {
        let clean = PhoneNumber.sanitizeInviteCode(code)
        guard clean.count == PhoneNumber.inviteCodeLength else { throw HeadstartError.badCode }
        return try await callables.acceptPair(code: clean).pairId
    }

    public func revoke() async throws {
        guard let pairId = pair?.id else { throw HeadstartError.notPaired }
        try await callables.revokePair(pairId: pairId)
    }

    // MARK: - Derived

    /// The other member's uid, or nil while the pair is pending.
    public func otherUid(selfUid: String) -> String? { pair?.other(than: selfUid) }

    /// Never nil — "Your partner" when the other side has no name yet.
    public func partnerName(selfUid: String) -> String {
        PartnerName.resolve(pair: pair, selfUid: selfUid)
    }
}
